File: backend/idempotency.py
"""Iter126b · Idempotency-Key middleware for Bucket-B write endpoints.

Contract
--------
Every Bucket-B POST (create / payment / upload / LR regenerate) that carries a
client-generated `Idempotency-Key` header will:

  1. Execute the underlying handler **exactly once** per unique
     (user_id, active company_id, method, path, key) tuple.
  2. Cache the resulting status code + body + preserved headers for 24 h
     (TTL index on `idempotency_keys.created_at`).
  3. Replay the cached response verbatim on any subsequent request with the
     same tuple — no duplicate Trip / Customer / Invoice / Payment / Upload
     row is ever created.
  4. Single-flight concurrent duplicates by claiming the composite `_id`
     with `insert_one`; losers poll until the winner completes (max 30 s).
  5. Isolate keys per user AND per active company — a key reused from a
     different user or company cannot replay another user's response.

Deliberately NOT covered (see PRD Iter126 safety matrix):
  · Bucket A endpoints — client Axios retry already handles those (no
    duplicate side-effects since they're naturally idempotent).
  · Bucket C endpoints — communications / LLM / bulk mutations / auth
    handshakes. These must never auto-retry, so we intentionally do not
    replay them; if a caller sends an Idempotency-Key on a Bucket-C route
    the middleware simply ignores it and passes through.

Failure / edge policy:
  · No Idempotency-Key header  → passthrough (backward compatible).
  · Malformed key (< 8 or > 128 chars) → passthrough (endpoint sees raw
    request; we do not surface a new 400 that older clients wouldn't).
  · Unauthenticated request → passthrough (endpoint returns its own 401;
    we never write an idempotency row for an unauthenticated caller).
  · Handler raises → claim row is deleted so a client retry can proceed.
  · 5xx response → NOT cached (transient error; client is expected to
    retry with the same key — Bucket-A retry logic + this middleware
    together achieve the "retry after 502/503 returns the eventual
    successful result" guarantee).
  · Response body > 5 MiB → NOT cached (avoid Mongo bloat); handler will
    re-run on retry. The three A→B LR endpoints back-fill `lr_number`
    deterministically so re-run still won't create duplicates.
"""
from __future__ import annotations

import asyncio
import base64
import hashlib
import logging
import re
from datetime import datetime, timezone

from fastapi import Request
from starlette.responses import Response

from db import db

logger = logging.getLogger(__name__)

# --------------------------------------------------------------------------- #
# Bucket B endpoint whitelist (locked with user 2026-Feb).
# --------------------------------------------------------------------------- #
BUCKET_B_PATTERNS: list[re.Pattern] = [re.compile(p) for p in (
    # top-level create endpoints
    r"^/api/trips$",
    r"^/api/companies$",
    r"^/api/customers$",
    r"^/api/suppliers$",
    r"^/api/drivers$",
    r"^/api/vehicles$",
    r"^/api/invoices$",
    r"^/api/products$",
    r"^/api/parties$",
    r"^/api/team$",
    r"^/api/templates$",
    r"^/api/fuel$",
    r"^/api/expenditure-types$",
    r"^/api/saved-trip-filters$",
    r"^/api/driver-shortage-policies$",
    # sub-resource creates
    r"^/api/customers/[^/]+/ship-sites$",
    r"^/api/customers/[^/]+/add-payment$",
    r"^/api/suppliers/[^/]+/payments$",
    r"^/api/invoices/[^/]+/payments$",
    r"^/api/drivers/[^/]+/ledger$",
    r"^/api/drivers/[^/]+/ledger/post-monthly-salary$",
    r"^/api/drivers/[^/]+/ledger/settle$",
    r"^/api/drivers/[^/]+/salary-masters$",
    r"^/api/trips/[^/]+/supplier-diesel$",
    r"^/api/trips/[^/]+/supplier-advance$",
    r"^/api/trips/[^/]+/field-override$",
    r"^/api/trips/[^/]+/duplicate$",
    r"^/api/trips/from-template/[^/]+$",
    r"^/api/trips/quick-repeat/[^/]+$",
    # uploads
    r"^/api/files/upload$",
    r"^/api/files/bulk-upload$",
    r"^/api/company/logo$",
    # bulk imports
    r"^/api/trips/import$",
    r"^/api/vehicles/bulk-import$",
    # A→B LR regenerate moves (audit + optional lr_number back-fill)
    r"^/api/trips/[^/]+/regenerate-lr$",
    r"^/api/trips/bulk-regenerate-lr$",
    r"^/api/trips/bulk-all-copies-zip$",
)]

MAX_STORED_BYTES = 5 * 1024 * 1024   # 5 MiB — bigger responses skip cache
POLL_INTERVAL_SEC = 0.15             # duplicate follower poll cadence
POLL_TIMEOUT_SEC = 30                # duplicate follower wall-clock cap
TTL_SECONDS = 24 * 3600              # 24 h replay window (user-approved)
KEY_MIN_LEN = 8
KEY_MAX_LEN = 128

# Response headers that must be preserved verbatim for a valid replay.
_PRESERVE_HEADERS = {
    "content-type",
    "content-disposition",
    # Iter125 bulk-all-copies-zip surfaces trip counts via response headers;
    # replay must not drop them or the frontend post-download modal breaks.
    "x-iter125-included",
    "x-iter125-skipped",
    "x-iter125-reasons",
    "x-iter125-pdfs",
}


def _matches_bucket_b(method: str, path: str) -> bool:
    if method.upper() != "POST":
        return False
    return any(rx.match(path) for rx in BUCKET_B_PATTERNS)


def _composite_id(user_id: str, company_id: str, method: str, path: str, key: str) -> str:
    payload = f"{user_id}|{company_id or '_'}|{method}|{path}|{key}".encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


async def _resolve_user_and_company(request: Request) -> tuple[str | None, str]:
    """Cheap Bearer-token lookup — mirrors auth.get_current_user() but read-only
    and returns (None, '') if the caller is not authenticated. We deliberately
    do NOT trigger the rolling-refresh write here; that stays in the handler.
    """
    token = request.cookies.get("session_token")
    if not token:
        auth = request.headers.get("Authorization", "")
        if auth.startswith("Bearer "):
            token = auth[7:]
    if not token:
        return None, ""
    try:
        session = await db.user_sessions.find_one(
            {"session_token": token}, {"_id": 0, "user_id": 1}
        )
    except Exception:
        return None, ""
    if not session:
        return None, ""
    return session.get("user_id"), request.headers.get("X-Company-Id", "") or ""


async def ensure_indexes() -> None:
    """Called at server startup — creates the 24h TTL index on
    idempotency_keys.created_at. Safe to call repeatedly."""
    try:
        await db.idempotency_keys.create_index(
            "created_at", expireAfterSeconds=TTL_SECONDS, name="ttl_created_at"
        )
    except Exception as e:
        logger.warning(f"idempotency_keys TTL index setup: {e}")


def _replay_response(row: dict) -> Response:
    body_b64 = row.get("response_body_b64") or ""
    body = base64.b64decode(body_b64) if body_b64 else b""
    headers = dict(row.get("response_headers") or {})
    headers["x-idempotent-replay"] = "1"
    return Response(
        content=body,
        status_code=int(row.get("status_code", 200)),
        headers=headers,
        media_type=row.get("media_type") or headers.get("content-type"),
    )


async def idempotency_middleware(request: Request, call_next):
    path = request.url.path or ""
    method = request.method
    key = (request.headers.get("Idempotency-Key") or "").strip()

    # Passthrough fast-path — anything outside the locked Bucket B whitelist
    # or without a valid client-supplied key is entirely unaffected.
    if not key or not _matches_bucket_b(method, path):
        return await call_next(request)
    if not (KEY_MIN_LEN <= len(key) <= KEY_MAX_LEN):
        return await call_next(request)

    user_id, company_id = await _resolve_user_and_company(request)
    if not user_id:
        return await call_next(request)

    kid = _composite_id(user_id, company_id, method, path, key)
    coll = db.idempotency_keys
    now = datetime.now(timezone.utc)

    # Try to claim the key (single-flight enforcement via unique _id).
    claimed = True
    try:
        await coll.insert_one({
            "_id": kid,
            "user_id": user_id,
            "company_id": company_id,
            "method": method,
            "path": path,
            "key": key,
            "status": "in_progress",
            "created_at": now,
        })
    except Exception:
        claimed = False  # duplicate _id — replay or wait

    if not claimed:
        # Follower path — either replay a completed row or poll until the
        # leader finishes. On timeout we fall through and let the request
        # run again (better than blocking a caller forever); the caller can
        # still send the same Idempotency-Key on a subsequent retry.
        row = await coll.find_one({"_id": kid})
        if row and row.get("status") == "completed":
            return _replay_response(row)
        deadline = asyncio.get_event_loop().time() + POLL_TIMEOUT_SEC
        while asyncio.get_event_loop().time() < deadline:
            await asyncio.sleep(POLL_INTERVAL_SEC)
            row = await coll.find_one({"_id": kid})
            if row and row.get("status") == "completed":
                return _replay_response(row)
        return await call_next(request)

    # Leader path — run the actual handler.
    try:
        response = await call_next(request)
    except Exception:
        # Release claim so a client retry can proceed cleanly.
        try:
            await coll.delete_one({"_id": kid})
        except Exception:
            pass
        raise

    # Buffer the body so we can both return it AND cache it.
    body_bytes = b""
    try:
        async for chunk in response.body_iterator:
            body_bytes += chunk
    except Exception:
        # If the response has no body_iterator (rare — e.g. a plain Response
        # already assembled), fall back to `.body`.
        body_bytes = getattr(response, "body", b"") or b""

    cacheable = (response.status_code < 500) and (len(body_bytes) <= MAX_STORED_BYTES)
    if cacheable:
        preserved = {
            h: v for h, v in response.headers.items()
            if h.lower() in _PRESERVE_HEADERS
        }
        try:
            await coll.update_one(
                {"_id": kid},
                {"$set": {
                    "status": "completed",
                    "status_code": int(response.status_code),
                    "response_body_b64": base64.b64encode(body_bytes).decode("ascii"),
                    "response_headers": preserved,
                    "media_type": response.media_type,
                    "completed_at": datetime.now(timezone.utc),
                }},
            )
        except Exception as e:
            logger.warning(f"idempotency cache write failed for {path}: {e}")
    else:
        # Not cached — drop the claim so a follow-up retry with the same key
        # can re-run and (for A→B LR endpoints) still be duplicate-safe.
        try:
            await coll.delete_one({"_id": kid})
        except Exception:
            pass

    # Return a fresh Response — the original body_iterator has been drained.
    return Response(
        content=body_bytes,
        status_code=response.status_code,
        headers=dict(response.headers),
        media_type=response.media_type,
    )

File: backend-node/src/idempotency.ts
import { createHash } from 'node:crypto';
import type { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import type { Db } from 'mongodb';

/**
 * TRUKVIA · Phase-3 · Gate-4 — Node Bucket-B idempotency port.
 * Faithful shadow of `backend/idempotency.py`. Scope is deliberately narrow:
 * only ONE route (`POST /api/saved-trip-filters`). Additional entries require
 * a separate authorised gate.
 *
 * Semantics (verbatim from Python):
 *   - Only POST + allowlisted paths + key len 8..128 activates.
 *   - Composite `_id` = sha256(user | company_or_'_' | METHOD | path | key).
 *   - Leader claims via insertOne({_id}); dup = follower.
 *   - Follower polls every 150ms up to 30s; if leader completes → replay.
 *   - handler throw → delete claim row.
 *   - cache only when status < 500 and body <= 5 MiB.
 *   - 5xx/oversize → delete claim (no cache).
 *   - Replay adds `x-idempotent-replay: 1`.
 *   - `ensureIdempotencyIndex` is safe/idempotent.
 */

const BUCKET_B_PATTERNS: readonly RegExp[] = [
  /^\/api\/saved-trip-filters$/,
];

const KEY_MIN_LEN = 8;
const KEY_MAX_LEN = 128;
const TTL_SECONDS = 24 * 3600;
const MAX_STORED_BYTES = 5 * 1024 * 1024;
const POLL_INTERVAL_MS = 150;
const POLL_TIMEOUT_MS = 30_000;

const PRESERVE_HEADERS = new Set<string>([
  'content-type',
  'content-disposition',
  'x-iter125-included',
  'x-iter125-skipped',
  'x-iter125-reasons',
  'x-iter125-pdfs',
]);

export function matchesBucketB(method: string, path: string): boolean {
  if (method.toUpperCase() !== 'POST') return false;
  return BUCKET_B_PATTERNS.some((rx) => rx.test(path));
}

export function compositeId(
  userId: string,
  companyId: string,
  method: string,
  path: string,
  key: string,
): string {
  const payload = `${userId}|${companyId || '_'}|${method}|${path}|${key}`;
  return createHash('sha256').update(payload, 'utf8').digest('hex');
}

interface SessionLite { user_id: string }

async function resolveUserAndCompany(req: FastifyRequest, db: Db): Promise<{ userId: string | null; companyId: string }> {
  let token: string | null = null;
  const cookie = req.headers['cookie'];
  if (typeof cookie === 'string') {
    for (const part of cookie.split(';')) {
      const [n, ...rest] = part.split('=');
      if (n?.trim() === 'session_token' && rest.length > 0) {
        const raw = rest.join('=').trim();
        if (raw) token = decodeURIComponent(raw);
        break;
      }
    }
  }
  if (!token) {
    const auth = req.headers['authorization'];
    if (typeof auth === 'string' && auth.startsWith('Bearer ')) {
      const t = auth.slice(7).trim();
      if (t) token = t;
    }
  }
  if (!token) return { userId: null, companyId: '' };
  try {
    const s = await db.collection<SessionLite>('user_sessions').findOne(
      { session_token: token }, { projection: { _id: 0, user_id: 1 } });
    if (!s) return { userId: null, companyId: '' };
    const cid = req.headers['x-company-id'];
    const first = Array.isArray(cid) ? cid[0] : cid;
    return { userId: s.user_id, companyId: typeof first === 'string' ? first : '' };
  } catch {
    return { userId: null, companyId: '' };
  }
}

interface ClaimRow {
  _id: string;
  status: 'in_progress' | 'completed';
  status_code?: number;
  response_body_b64?: string;
  response_headers?: Record<string, string>;
  media_type?: string;
}

export async function ensureIdempotencyIndex(db: Db): Promise<void> {
  try {
    await db.collection('idempotency_keys').createIndex(
      { created_at: 1 },
      { expireAfterSeconds: TTL_SECONDS, name: 'ttl_created_at' },
    );
  } catch {
    /* mirror Python: log-and-swallow; index may pre-exist */
  }
}

export function registerIdempotency(app: FastifyInstance, deps: { db: Db }): void {
  const { db } = deps;

  app.addHook('onRequest', async (req, reply) => {
    const path = req.url.split('?')[0] ?? '';
    const method = req.method;
    const key = (req.headers['idempotency-key'] as string | undefined ?? '').trim();

    if (!key || !matchesBucketB(method, path)) return;
    if (key.length < KEY_MIN_LEN || key.length > KEY_MAX_LEN) return;

    const { userId, companyId } = await resolveUserAndCompany(req, db);
    if (!userId) return;

    const kid = compositeId(userId, companyId, method, path, key);
    const coll = db.collection<ClaimRow>('idempotency_keys');
    const now = new Date();

    let claimed = true;
    try {
      await coll.insertOne({
        _id: kid, status: 'in_progress',
        // metadata for parity with Python's row shape
        ...(({ user_id: userId, company_id: companyId, method, path, key, created_at: now }) as unknown as object),
      } as unknown as ClaimRow);
    } catch {
      claimed = false;
    }

    if (!claimed) {
      // Follower path
      const deadline = Date.now() + POLL_TIMEOUT_MS;
      for (;;) {
        const row = await coll.findOne({ _id: kid });
        if (row && row.status === 'completed') {
          await sendReplay(reply, row);
          return;
        }
        if (Date.now() >= deadline) return; // fall through: run handler as fresh leader
        await new Promise((r) => setTimeout(r, POLL_INTERVAL_MS));
      }
    }

    // Leader path — attach onSend to capture and cache the response.
    (req as unknown as { __idem_kid?: string }).__idem_kid = kid;
  });

  app.addHook('onSend', async (req, reply, payload) => {
    const kid = (req as unknown as { __idem_kid?: string }).__idem_kid;
    if (!kid) return payload;

    const coll = db.collection<ClaimRow>('idempotency_keys');
    const status = reply.statusCode;

    let body: Buffer;
    if (Buffer.isBuffer(payload)) body = payload;
    else if (typeof payload === 'string') body = Buffer.from(payload, 'utf8');
    else if (payload == null) body = Buffer.alloc(0);
    else {
      // Streams / iterables are not cacheable in this minimal port
      try { await coll.deleteOne({ _id: kid }); } catch { /* noop */ }
      return payload;
    }

    const cacheable = status < 500 && body.length <= MAX_STORED_BYTES;
    if (!cacheable) {
      try { await coll.deleteOne({ _id: kid }); } catch { /* noop */ }
      return payload;
    }

    const preserved: Record<string, string> = {};
    for (const [h, v] of Object.entries(reply.getHeaders())) {
      if (typeof v === 'string' && PRESERVE_HEADERS.has(h.toLowerCase())) preserved[h.toLowerCase()] = v;
    }
    try {
      await coll.updateOne(
        { _id: kid },
        { $set: {
          status: 'completed',
          status_code: status,
          response_body_b64: body.toString('base64'),
          response_headers: preserved,
          media_type: preserved['content-type']?.split(';')[0]?.trim() ?? '',
          completed_at: new Date(),
        } },
      );
    } catch { /* mirror Python: log-and-swallow */ }
    return payload;
  });

  // Handler threw → also clear claim so a retry can proceed.
  app.addHook('onError', async (req) => {
    const kid = (req as unknown as { __idem_kid?: string }).__idem_kid;
    if (!kid) return;
    try { await db.collection<ClaimRow>('idempotency_keys').deleteOne({ _id: kid }); } catch { /* noop */ }
  });
}

async function sendReplay(reply: FastifyReply, row: ClaimRow): Promise<void> {
  const status = row.status_code ?? 200;
  const headers = row.response_headers ?? {};
  for (const [h, v] of Object.entries(headers)) reply.header(h, v);
  reply.header('x-idempotent-replay', '1');
  reply.code(status);
  const body = row.response_body_b64 ? Buffer.from(row.response_body_b64, 'base64') : Buffer.alloc(0);
  await reply.send(body);
}
